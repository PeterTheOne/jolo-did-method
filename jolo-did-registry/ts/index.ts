import RegistryContract from "jolocom-registry-contract";
import { IpfsStorageAgent } from "./ipfs";
import { IDidDocument } from "@decentralized-identity/did-common-typescript"

const JOLOCOM_PUBLIC_PROFILE_TYPE = "JolocomPublicProfile"
export const infura =  'https://rinkeby.infura.io/v3/64fa85ca0b28483ea90919a83630d5d8'
export const jolocomContract = '0xd4351c3f383d79ba378ed1875275b1e7b960f120'
export const jolocomIpfsHost = 'https://ipfs.jolocom.com:443'

export function getRegistry(providerUrl: string = infura, contractAddress: string = jolocomContract, ipfsHost: string = jolocomIpfsHost) {
  const registryContract = new RegistryContract(contractAddress, providerUrl)
  const ipfs = new IpfsStorageAgent(ipfsHost)
  return {
    commitDidDoc: async (privateKeyHex: string, didDocument: IDidDocument, publicProfile?: any): Promise<IDidDocument> => {
      let publicProfileSection, profileServiceIndex
      if (didDocument.service)
        profileServiceIndex = didDocument.service.findIndex(s => s.type === JOLOCOM_PUBLIC_PROFILE_TYPE)

      if (publicProfile) {
        const profileHash = await ipfs.storeJSON(publicProfile)
        publicProfileSection = generatePublicProfileServiceSection(didDocument.id, profileHash)
        if (!didDocument.service)
          didDocument.service = []
        if (profileServiceIndex === -1 || profileServiceIndex === undefined)
          didDocument.service.push(publicProfileSection)
        else
          didDocument.service[profileServiceIndex] = publicProfileSection
      } else {
        if (profileServiceIndex > -1)
          didDocument.service?.splice(profileServiceIndex, 1)
      }

      const documentHash = await ipfs.storeJSON(didDocument)

      await registryContract.updateDID(Buffer.from(privateKeyHex, 'hex'), didDocument.id, documentHash)
      return didDocument
    }
  }
}


/**
 * Instantiates the {@link ServiceEndpointsSection} class based on passed arguments
 * @param did - The did of the did document owner
 * @param profileIpfsHash - IPFS hash that can be used to dereference the public profile credential
 * @internal
 */

function generatePublicProfileServiceSection(
  did: string,
  profileIpfsHash: string,
) {
  return {
    id: `${ did };jolocomPubProfile`,
    serviceEndpoint: `ipfs://${ profileIpfsHash }`,
    description: 'Verifiable Credential describing entity profile',
    type: JOLOCOM_PUBLIC_PROFILE_TYPE,
  }
}
