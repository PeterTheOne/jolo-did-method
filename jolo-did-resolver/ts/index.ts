import { DIDDocument, ParsedDID, Resolver } from "did-resolver";
import EthereumResolver from "jolocom-registry-contract";
import { IpfsStorageAgent } from "./ipfs";

const CONTRACT_ADDRESS = '0xd4351c3f383d79ba378ed1875275b1e7b960f120';
const PROVIDER_URI = 'https://rinkeby.infura.io/';
const IPFS_ENDPOINT = 'https://ipfs.jolocom.com:443'

export function getResolver(providerUri: string = PROVIDER_URI, contractAddress: string = CONTRACT_ADDRESS, ipfsHost: string = IPFS_ENDPOINT) {
  const ethereumConnector = new EthereumResolver(contractAddress, providerUri)
  const ipfsAgent = new IpfsStorageAgent(ipfsHost)

  async function resolve(
    did: string,
    parsed: ParsedDID,
    didResolver: Resolver
  ): Promise<DIDDocument | null> {
    const ipfsHash = await ethereumConnector.resolveDID(parsed.id);

    return (await ipfsAgent.catJSON(ipfsHash)) as DIDDocument
  }

  return { "jolo": resolve }
}

export async function getPublicProfile(ipfsHash: string, ipfsHost: string): Promise<any> {
  const ipfsAgent = new  IpfsStorageAgent(ipfsHost)
  return ipfsAgent.catJSON(ipfsHash)
}