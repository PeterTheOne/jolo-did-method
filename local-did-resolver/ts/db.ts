import { extractMessageId } from "."

interface DbState {
  [did: string]: string[]
}

// SMALL MOCK IMPLEMENTATION, this will mutate the initialState
export const createDb = (initialState: DbState = {}): InternalDb =>
  ({
    append: async (event: string) => {
      const eventId = extractMessageId(event)
      if (!initialState[eventId].length) {
        initialState[eventId] = []
      }

      initialState[eventId].push(event)
      return true
    },
    delete: async (id: string) => {
      initialState[id] = [] 
      return true
    },
    read: async id => initialState[id]
  })

export interface InternalDb {
  read: (id: string) => Promise<string[]>
  append: (event: string) => Promise<boolean>
  delete: (id: string) => Promise<boolean>
}
