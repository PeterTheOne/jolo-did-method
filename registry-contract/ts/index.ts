import { Contract, ethers } from "ethers";
import { BaseProvider } from "ethers/providers";

const RegistryContract = require('../build/contracts/Registry.json')

export default class EthereumConnector {
  private readonly provider: BaseProvider
  private contract: Contract

  constructor(address: string, providerUri: string) {
    this.provider = new ethers.providers.JsonRpcProvider(providerUri)
    this.contract = new ethers.Contract(address, RegistryContract.abi, this.provider)
  }

  async resolveDID(did: string): Promise<string> {
    const keyHash = stripMethodPrefix(did)
    return await this.contract.getRecord(keyHash)
  }

  async updateDIDRecord(ethereumKey: any, did: string, newHash: string): Promise<void> {
    const keyHash = stripMethodPrefix(did)

    const wallet = new ethers.Wallet(ethereumKey, this.provider)
    const signer = this.contract.connect(wallet)
    const tx = await signer.setRecord(keyHash, newHash)
    await tx.wait()
  }
}

function stripMethodPrefix(did: string) {
  return `0x${ did.substring(did.lastIndexOf(':') + 1) }`
}