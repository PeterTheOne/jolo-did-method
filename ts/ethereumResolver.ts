import * as Web3 from 'web3'
import * as RegistryContract from '../build/contracts/Registry.json'

export default class EthereumResolver{
  constructor() {
    const address = '0xc4b48901af7891d83ce83877e1f8fb4c81a94907'

    this.web3 = new Web3(new Web3.providers.HttpProvider('http://localhost:8545'))
    this.indexContract = new this.web3.eth.Contract(RegistryContract.abi, address)
  }
  
  resolveDID(did: string): Promise<string> {
    return new Promise((resolve, reject) => {
      const keyHash = this._stripMethodPrefix(did)
      this.indexContract.methods.getRecord(keyHash).call((error, result) => {
        if (error) {
          return reject(error)
        }
        return resolve(result)
      })
    })
  }

  updateDIDRecord(sender: string, did: string, newHash: string): Promise<string> {
    return new Promise((resolve, reject) => {
      const keyHash = this._stripMethodPrefix(did)
      

      this.indexContract.methods.setRecord(keyHash, newHash).send({
        from: sender,
      },(error, result) => {
        if (error) {
          return reject(error)
        }

        /* The web3 provider needs to support WSS, 
         * will be configured at later stage

        this.indexContract.events.registrationSuccess((error, result) => {
          console.log(error)
          console.log(result)
        })

        */

        return resolve()
      })
    })
  }

  private _stripMethodPrefix(did: string): string {
    return did.substring(did.lastIndexOf(':') + 1)
  }
}
