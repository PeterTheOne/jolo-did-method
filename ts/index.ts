import * as wallet from 'ethereumjs-wallet'
import * as Transaction from 'ethereumjs-tx'

const RegistryContract = require('../build/contracts/Registry.json')
const Web3 = require('web3')

export default class EthereumResolver{
  private web3: any
  private indexContract: any

  constructor(address: string, providerUri: string) {
    const provider =  new Web3.providers.HttpProvider(providerUri)
    this.web3 = new Web3(provider)
    this.contractAddress = address
    this.indexContract = new this.web3.eth.Contract(RegistryContract.abi, address)
  }
  
  resolveDID(did: string): Promise<string> {
    return new Promise((resolve, reject) => {
      const keyHash = this._stripMethodPrefix(did)
      this.indexContract.methods.getRecord(keyHash).call((error, result) => {
        if (error) {
          return reject(error)
        }
        return resolve(result)
      })
    })
  }

  updateDIDRecord(ethereumKey: any, did: string, newHash: string): Promise<string> {
    const gasLimit = 100000
    const gasPrice = 20e9

    const w = wallet.fromPrivateKey(ethereumKey)
    const address = w.getAddress().toString('hex')
    const keyHash = this._stripMethodPrefix(did)
    
    const callData = this.indexContract.methods.setRecord(keyHash, newHash)
      .encodeABI()

    return this.web3.eth.getTransactionCount(address).then(nonce => {
      const tx = new Transaction({
        nonce: nonce,
        gasLimit,
        gasPrice,
        data: callData,
        to: this.contractAddress
      })

      tx.sign(ethereumKey)
      const serializedTx = tx.serialize()

      return new Promise((resolve, reject) => {
        this.web3.eth.sendSignedTransaction(`0x${serializedTx.toString('hex')}`)
        .on('confirmation', () => resolve())
        .on('error', (err) => reject(err))
      })
    })
  }

  private _stripMethodPrefix(did: string): string {
    return did.substring(did.lastIndexOf(':') + 1)
  }
}
