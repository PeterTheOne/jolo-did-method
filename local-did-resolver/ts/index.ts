import { DIDDocument, ParsedDID, Resolver } from "did-resolver";
import { InternalDb } from "./db";
import { validateEvents, getIcp } from '@jolocom/native-utils-node'

export function getResolver(dbInstance: InternalDb) {
  return { 
    un: async (did: string, parsed: ParsedDID, didResolver: Resolver): Promise<DIDDocument | null> => {
      const events = await dbInstance.read(did)

      if (events && events.length) {
        return validateEvents(JSON.stringify(events)) // TODO
        .then(JSON.parse)
      }

      return null
    }
  }
}

export function getRegistrar(dbInstance: InternalDb) {
  return {
    update: async (events: string[]) => {
      try {
        const keyEventId = extractMessageId(events[0])
        const previousEvents = await dbInstance.read(keyEventId) || []

        const document = await validateEvents(
          previousEvents.concat(events)
        )

        const parsedDidDocument = JSON.parse(document)
        dbInstance.append(events)
        return parsedDidDocument
        } catch (err) {
          return err
      }
    },
    delete: (id: string): Promise<boolean> => dbInstance.delete(id),
    create: () => {
      console.warn('For testing')
      return getIcp() // TODO the args are not needed anymore
    }
  }
}

export const extractMessageId = (serializedEvent: string): string => {
  const { keyEvent } = separateEventAndSignature(serializedEvent)
  try {
    return `did:un:${JSON.parse(keyEvent).id}`
  } catch(err) {
    return err
  }
}

const separateEventAndSignature = (serializedEvent: string): {keyEvent: string, signatures: string[]} => {
  const separator = '\r\n\r\n'
  const signatureSeparator = '\n'

  const [keyEvent, signatureBlock] = serializedEvent.split(separator)
  const signatures = signatureBlock.split(signatureSeparator)

  return { keyEvent, signatures }
}
