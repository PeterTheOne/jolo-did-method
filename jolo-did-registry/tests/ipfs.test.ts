import { IpfsStorageAgent, jolocomIpfsStorageAgent } from "../ts/ipfs";

describe('IPFS Agent', () => {
  it('should store file on ipfs', async () => {
    const ipfs = new IpfsStorageAgent('http://127.0.0.1:5001')
    // ipfs.fetchImplementation = jest.fn()
    const resutl  = await ipfs.storeJSON({ data: 'test' }, true)
    console.log(resutl)
    // expect(ipfs.fetchImplementation.mock.calls[0]).toMatchSnapshot()
  });

  it('should store file on ipfs really', async () => {
    const ipfs = new IpfsStorageAgent('http://127.0.0.1:5001')
    // ipfs.fetchImplementation = jest.fn()
    const result  = await ipfs.storeJSON({ data: 'test' }, true)
    console.log(result)
    // expect(ipfs.fetchImplementation).toMatchSnapshot()
  });
})