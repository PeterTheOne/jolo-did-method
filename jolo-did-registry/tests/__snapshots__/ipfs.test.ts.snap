// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IPFS Agent should store file on ipfs 1`] = `
Array [
  "test/api/v0/add?pin=true",
  Object {
    "body": FormData {
      "_boundary": "--------------------------119784020157390999052484",
      "_currentStream": null,
      "_insideLoop": false,
      "_overheadLength": 143,
      "_pendingNext": false,
      "_released": false,
      "_streams": Array [
        "----------------------------119784020157390999052484
Content-Disposition: form-data; name=\\"file\\"
Content-Type: application/octet-stream

",
        Object {
          "data": Array [
            123,
            34,
            100,
            97,
            116,
            97,
            34,
            58,
            34,
            116,
            101,
            115,
            116,
            34,
            125,
          ],
          "type": "Buffer",
        },
        [Function],
      ],
      "_valueLength": 15,
      "_valuesToMeasure": Array [],
      "dataSize": 0,
      "maxDataSize": 2097152,
      "pauseStreams": true,
      "readable": true,
      "writable": false,
    },
    "method": "POST",
  },
]
`;
